flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalDriver.java

```diff
@@ -122,6 +125,19 @@ public void close() throws Exception {
         client.getConnectionStateListenable().removeListener(connectionStateListener);
 
         cache.close();
+
+        try {
+            if (client.getZookeeperClient().isConnected()
+                    && !connectionInformationPath.contains(RESOURCE_MANAGER_NODE)) {
```

Comment:
@KarmaGYZ Sorry for jumping in that late. I was on vacation for the past three weeks. I have a question about this change: Why do we have to filter the ResourceManager node here? It feels odd to refer to the ResourceManager in the `ZooKeeperLeaderRetrievalDriver`. Knowledge about the ResourceManager component should be out-of-scope for the driver implementation. WDYT?

Reply:
I agree that this is a rather tricky fix. As Tison mentioned, JobMaster and Dispatcher will use the same curator framework at the JobManager side. Close the watch on resource manager when JobMaster exit will influence other JobMasters and Dispatcher, thus I arbitrarily exclude the RM here. We can also construct an exclusive curator framework for each JobMaster, but I think a cleaner solution should be implemented within curator or zookeeper, then we should remove this fix. WDYT?

Reply:
Also an exclusive curator framework will definitely harm the performance and add additional pressure to ZooKeeper.